File: Data Structures & Algorithms/daily-temperatures/submission-1.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        from collections import deque

        stack = deque([])

        res = [0] * len(temperatures)

        for i, temp in enumerate(temperatures):
            if not stack:
                stack.append((temp, i))
            else:
                while stack and temp > stack[-1][0]:
                    res[stack[-1][1]] = i - stack[-1][1]
                    stack.pop()
                stack.append((temp, i))

        return res

File: Data Structures & Algorithms/generate-parentheses/submission-1.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        Stack<String> stack = new Stack<>();
        backTracking(0, 0, n, res, stack);
        return res;
    }

    public void backTracking(int openN, int closeN, int n, List<String> res, Stack<String> stack){
        if(openN == n && closeN == n){
            StringBuilder sb = new StringBuilder();
            for(String s : stack){
                sb.append(s);        
            }
            res.add(sb.toString());
            return;
        }

        if(openN < n){
            stack.push("(");
            backTracking(openN + 1, closeN, n, res, stack);
            stack.pop();
        }

        if(closeN < openN) {
            stack.push(")");
            backTracking(openN, closeN + 1, n, res, stack);
            stack.pop();
        }
    }

    
}
